xray/grpc_test.go

```diff
@@ -0,0 +1,260 @@
+	"github.com/stretchr/testify/assert"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+	"google.golang.org/grpc/test/bufconn"
+)
+
+type testGRPCPingService struct {
+	counter int32
+	mut     sync.Mutex
+
+	pb.TestServiceServer
+}
+
+func (s *testGRPCPingService) Ping(_ context.Context, req *pb.PingRequest) (*pb.PingResponse, error) {
+	time.Sleep(time.Duration(req.SleepTimeMs) * time.Millisecond)
+
+	s.mut.Lock()
+	s.counter++
+	counter := s.counter
+	s.mut.Unlock()
+
+	return &pb.PingResponse{
+		Value:   req.Value,
+		Counter: counter,
+	}, nil
+}
+
+func (s *testGRPCPingService) PingError(_ context.Context, req *pb.PingRequest) (*pb.Empty, error) {
+	code := codes.Code(req.ErrorCodeReturned)
+	return nil, status.Errorf(code, "Userspace error.")
+}
+
+func newGrpcServer(t *testing.T, opts ...grpc.ServerOption) *bufconn.Listener {
+	const bufSize = 1024 * 1024
+	lis := bufconn.Listen(bufSize)
+
+	s := grpc.NewServer(opts...)
+	pb.RegisterTestServiceServer(s, &testGRPCPingService{})
+	go func() {
+		if err := s.Serve(lis); err != nil {
+			t.Fatal(err)
+		}
+	}()
+
+	return lis
+}
+
+func newGrpcClient(t *testing.T, ctx context.Context, lis *bufconn.Listener, opts ...grpc.DialOption) (client pb.TestServiceClient, closeFunc func()) {
+	var bufDialer = func(ctx context.Context, address string) (net.Conn, error) {
+		return lis.Dial()
+	}
+
+	opts = append(opts, grpc.WithContextDialer(bufDialer), grpc.WithInsecure())
+
+	conn, err := grpc.DialContext(
+		ctx,
+		"bufnet",
+		opts...,
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+	closeFunc = func() {
+		if err := conn.Close(); err != nil {
+			t.Fatal(err)
+		}
+	}
+	client = pb.NewTestServiceClient(conn)
+	return
+}
+
+func TestGrpcUnaryClientInterceptor(t *testing.T) {
+	lis := newGrpcServer(
+		t,
+		grpc_middleware.WithUnaryServerChain(
+			UnaryServerInterceptor(context.Background(), NewFixedSegmentNamer("test")),
+		),
+	)
+	client, closeFunc := newGrpcClient(
+		t,
+		context.Background(),
+		lis,
+		grpc.WithUnaryInterceptor(UnaryClientInterceptor("bufnet")),
+	)
+	defer closeFunc()
+
+	t.Run("success response", func(t *testing.T) {
+		ctx, td := NewTestDaemon()
+		defer td.Close()
+
+		ctx2, root := BeginSegment(ctx, "Test")
+		_, err := client.Ping(
+			ctx2,
+			&pb.PingRequest{Value: "something", SleepTimeMs: 9999},
+		)
+		root.Close(nil)
+		if !assert.NoError(t, err) {
+			return
+		}
+
+		seg, err := td.Recv()
+		if !assert.NoError(t, err) {
+			return
+		}
+
+		var subseg *Segment
+		assert.NoError(t, json.Unmarshal(seg.Subsegments[0], &subseg))
+		assert.Equal(t, "remote", subseg.Namespace)
+		assert.Equal(t, "grpc://bufnet/mwitkow.testproto.TestService/Ping", subseg.HTTP.Request.URL)
+		assert.Equal(t, false, subseg.HTTP.Request.XForwardedFor)
+		assert.False(t, subseg.Throttle)
+		assert.False(t, subseg.Error)
+		assert.False(t, subseg.Fault)
+	})
+
+	t.Run("error response", func(t *testing.T) {
```

Comment:
In sync with above comments we should add cases to check `throttle` and `faults` values as well.